File: js/src/maplestory/api/dto/union/unionRaiderBlockDto.ts
import { UnionRaiderBlockPositionDto } from './unionRaiderBlockPositionDto';
import { UnionRaiderBlockControlPointDto } from './unionRaiderControlPointDto';
import { UnionRaiderBlockBody } from '../../response/union/unionRaiderBody';

/**
 * 유니온 블록 정보
 */
class UnionRaiderBlockDto {
  /**
   * 블록 모양 (전사, 마법사, 궁수, 도적, 해적, 메이플m, 하이브리드)
   */
  blockType: string;

  /**
   * 블록 해당 캐릭터 직업
   */
  blockClass: string;

  /**
   * 블록 해당 캐릭터 레벨
   */
  blockLevel: string;

  /**
   * 블록 기준점 좌표
   */
  blockControlPoint: UnionRaiderBlockControlPointDto;

  /**
   * 블록이 차지하고 있는 영역 좌표들 (null:미 배치 시)
   */
  blockPosition: UnionRaiderBlockPositionDto[] | null;

  constructor(obj: UnionRaiderBlockBody) {
    const {
      block_type,
      block_class,
      block_level,
      block_control_point,
      block_position,
    } = obj;

    this.blockType = block_type;
    this.blockClass = block_class;
    this.blockLevel = block_level;
    this.blockControlPoint = new UnionRaiderBlockControlPointDto(
      block_control_point,
    );
    this.blockPosition = block_position
      ? block_position.map(
          (position) => new UnionRaiderBlockPositionDto(position),
        )
      : null;
  }
}

export { UnionRaiderBlockDto };

File: java/src/main/java/dev/spiralmoon/maplestory/api/tms/dto/union/UnionRaiderBlockDTO.java
package dev.spiralmoon.maplestory.api.tms.dto.union;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.util.List;

/**
 * 聯盟方塊資訊
 */
@AllArgsConstructor
@NoArgsConstructor
@Data
@ToString
public class UnionRaiderBlockDTO implements dev.spiralmoon.maplestory.api.common.dto.union.UnionRaiderBlockDTO<UnionRaiderBlockControlPointDTO, UnionRaiderBlockPositionDTO> {

    /**
     * 方塊編制 (戰士、法師、弓箭手、盜賊、海盜、混合)
     */
    @JsonProperty("block_type")
    private String blockType;

    /**
     * 方塊角色職業
     */
    @JsonProperty("block_class")
    private String blockClass;

    /**
     * 方塊角色等級
     */
    @JsonProperty("block_level")
    private String blockLevel;

    /**
     * 方塊基準點座標
     */
    @JsonProperty("block_control_point")
    private UnionRaiderBlockControlPointDTO blockControlPoint;

    /**
     * 方塊佔領區域的座標 (null：未部署時)
     */
    @JsonProperty("block_position")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<UnionRaiderBlockPositionDTO> blockPosition;
}
